reward/staking_info_db.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2020 The klaytn Authors
+// This file is part of the klaytn library.
+//
+// The klaytn library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The klaytn library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the klaytn library. If not, see <http://www.gnu.org/licenses/>.
+
+package reward
+
+import (
+	"encoding/binary"
+	"encoding/json"
+	"github.com/klaytn/klaytn/storage/database"
+)
+
+type stakingInfoDB struct {
+	db database.Database
+}
+
+func (sdb *stakingInfoDB) get(blockNum uint64) *StakingInfo {
+	if sdb.db == nil {
+		logger.Debug("stakingInfoDB.get() is called but stakingInfoDB is not set.")
+		return nil
+	}
+
+	key := intToByte(blockNum)
+	value, err := sdb.db.Get(key)
+	if err != nil {
+		logger.Error("Failed to get staking info from stakingInfoDB.", "blockNum", blockNum, "err", err)
+		return nil
+	}
+
+	stakingInfo := new(StakingInfo)
+	err = json.Unmarshal(value, stakingInfo)
+	if err != nil {
+		logger.Error("Failed to unmarshal staking info.", "blockNum", blockNum, "err", err, "raw staking info", value)
+		return nil
+	}
+
+	return stakingInfo
+}
+
+func (sdb *stakingInfoDB) add(stakingInfo *StakingInfo) {
+	if sdb.db == nil {
+		logger.Debug("stakingInfoDB.add() is called but stakingInfoDB is not set.")
+		return
+	}
+	key := intToByte(stakingInfo.BlockNum)
+
+	has, _ := sdb.db.Has(key)
```

Comment:
Why don't you skip the check `Has`?

Reply:
I put it to make the logic same with cache.  Also, I was afraid that something might go wrong and a new value is stored in DB.
However, I couldn't find such case when I was testing. I will remove `Has` since it will take more time. I will also add a couple of logger to check just in case.